examples/runners/ash/src/main.rs

```diff
@@ -1,259 +1,229 @@
-    message_severity: vk::DebugUtilsMessageSeverityFlagsEXT,
-    message_type: vk::DebugUtilsMessageTypeFlagsEXT,
-    p_callback_data: *const vk::DebugUtilsMessengerCallbackDataEXT,
-    _user_data: *mut std::os::raw::c_void,
-) -> vk::Bool32 {
-    let callback_data = *p_callback_data;
-    let message_id_number: i32 = callback_data.message_id_number as i32;
-
-    let message_id_name = if callback_data.p_message_id_name.is_null() {
-        Cow::from("")
-    } else {
-        CStr::from_ptr(callback_data.p_message_id_name).to_string_lossy()
-    };
+    ctx.build_pipelines(
+        vk::PipelineCache::null(),
+        vec![(
+            VertexShaderEntryPoint {
+                module: "sky_shader".into(),
+                entry_point: "main_vs".into(),
+            },
+            FragmentShaderEntryPoint {
+                module: "sky_shader".into(),
+                entry_point: "main_fs".into(),
+            },
+        )],
+    );
 
-    let message = if callback_data.p_message.is_null() {
-        Cow::from("")
-    } else {
-        CStr::from_ptr(callback_data.p_message).to_string_lossy()
-    };
+    event_loop.run(move |event, _window_target, control_flow| match event {
+        Event::RedrawEventsCleared { .. } => {
+            if !IS_COMPILING.load(Ordering::SeqCst) && NEEDS_REBUILD.load(Ordering::SeqCst) {
+                // if a recompile isn't in progress, this is the only thread.
+                unsafe {
+                    for SpirvShader { name, spirv } in NEW_SHADERS.drain(..) {
+                        ctx.insert_shader_module(name, spirv);
+                    }
+                }
+                ctx.rebuild_pipelines(vk::PipelineCache::null());
+                NEEDS_REBUILD.store(false, Ordering::SeqCst);
+            }
+            ctx.render();
+        }
+        Event::WindowEvent { event, .. } => match event {
+            WindowEvent::KeyboardInput { input, .. } => match input.virtual_keycode {
+                Some(VirtualKeyCode::Escape) => *control_flow = ControlFlow::Exit,
+                Some(VirtualKeyCode::F5) => {
+                    // cannot start multiple recompiles at once, cannot cancel either
+                    if !IS_COMPILING.compare_and_swap(false, true, Ordering::SeqCst) {
+                        std::thread::spawn(|| {
+                            // IS_COMPILING is set, main thread will not read this
+                            unsafe {
+                                NEW_SHADERS = compile_shaders();
+                            }
+                            NEEDS_REBUILD.store(true, Ordering::SeqCst);
+                            IS_COMPILING.store(false, Ordering::SeqCst);
+                        });
+                    }
+                    *control_flow = ControlFlow::Wait;
+                }
+                _ => *control_flow = ControlFlow::Wait,
+            },
+            WindowEvent::Resized(_) => {
+                ctx.recreate_swapchain();
```

Comment:
It looks like you're not resizing the swapchain (at least not passing through the new window resolution here)?

There should probably also be a condition around resizing down to a zero sized swapchain, this can happen at least on Windows when minimizing the window, but also when resizing the window to be really tiny.

Reply:
Both issues here have been addressed.  Viewports and scissors are now recreated when the window is resized, and rendering is paused when width or height is 0.  Resolution still isn't passed in here though, the window is stored in the RenderCtx so the resolution is retrieved via a method on that.

Unrelated note about this same code location: `recreate_swapchain` is done both here as well as in response to the out of date error returns during rendering.  On Windows, either of these is sufficient but this one takes precedence and produces nicer looking results (no black bars around the rendered portion on the frame drawn during a resize).